core/app/models/spree/promotion_code/batch_builder.rb

```diff
@@ -0,0 +1,66 @@
+class ::Spree::PromotionCode::BatchBuilder
+  attr_reader :promotion_code_batch
+  delegate :promotion, :number_of_codes, :base_code, to: :promotion_code_batch
+
+  class_attribute :random_code_length, :batch_size, :sample_characters
+  self.random_code_length = 6
+  self.batch_size = 1_000
+  self.sample_characters = ('a'..'z').to_a + (2..9).to_a.map(&:to_s)
+
+  def initialize(promotion_code_batch)
+    @promotion_code_batch = promotion_code_batch
+  end
+
+  def build_promotion_codes
+    generate_random_codes
+
+    Spree::PromotionCodeBatchMailer
+      .promotion_code_batch_finished(promotion_code_batch)
+      .deliver_now
+  rescue => e
```

Comment:
Moving the email to the Job makes sense to me but I think setting the error on the promo_code_batch should stay here.  We allow a row to exist in the `spree_promotion_code_batches` table when it may have failed part way through, so marking it with an error makes sense to me, and this seems like the place to do it.